Auth/Auth.Platform.iOS/Authenticator/OIDCAuthenticator.cs

```diff
@@ -0,0 +1,108 @@
+
+        private TaskCompletionSource<User> authenticateTaskComplete;
+        private Task<User> authenticateTask;
+
+        private IHttpServiceModule httpService;
+
+        public OIDCAuthenticator(AuthenticationConfig authenticationConfig, KeycloakConfig keycloakConfig, ICredentialManager credentialManager, IHttpServiceModule httpServiceModule)
+        {
+            this.authenticationConfig = nonNull(authenticationConfig, "authenticationConfig");
+            this.keycloakConfig = nonNull(keycloakConfig, "keycloakConfig");
+            this.credentialManager = nonNull(credentialManager, "credentialManager");
+            this.httpService = nonNull(httpServiceModule, "httpServiceModule");
+        }
+
+        public Task<User> Authenticate(IAuthenticateOptions authenticateOptions)
+        {
+            IOSAuthenticateOptions options = nonNull((IOSAuthenticateOptions)authenticateOptions, "authenticateOptions");
+            ServiceConfiguration oidServiceConfiguration = new ServiceConfiguration(this.keycloakConfig.AuthenticationEndpoint, this.keycloakConfig.TokenEndpoint); // = GetAuthorizationServiceConfiguration();
+            AuthorizationRequest oidAuthRequest = new AuthorizationRequest(oidServiceConfiguration,
+                                                                           this.keycloakConfig.ResourceId,
+                                                                           new String[] { Scope.OpenId, Scope.Profile },
+                                                                           this.authenticationConfig.RedirectUri,
+                                                                           ResponseType.Code, 
+                                                                           null);
+
+            authenticateTaskComplete = new TaskCompletionSource<User>();
+            authenticateTask = authenticateTaskComplete.Task;
+
+            this.currentAuthorizationFlow = startAuthorizationFlow(oidAuthRequest, options.PresentingViewController, (OIDCCredential credential, NSError error) =>
+            {
+                if (credential != null) {
+                    authenticateTaskComplete.TrySetResult(User.NewUser().FromUnverifiedCredential(credential, this.keycloakConfig.ResourceId));
+                } else {
+                    authenticateTaskComplete.TrySetException(new Exception(error.LocalizedDescription));
+                }
+            });
+
+            return authenticateTask;
+        }
+
+        public async Task<bool> Logout(User currentUser)
+        {
+            nonNull(currentUser, "current user");
+            string identityToken = currentUser.IdentityToken;
+            var logoutUrl = keycloakConfig.LogoutUrl(identityToken, authenticationConfig.RedirectUri.ToString());
+
+            TaskCompletionSource<bool> completionSource = new TaskCompletionSource<bool>();
+            IHttpResponse response = await httpService.NewRequest().Get(logoutUrl).Execute();
```

Comment:
using `var` may be better in this case I think. It's already called response, so you know what to expect in the code.

Reply:
@wojta I don't think that the returned type from `Execute` is so evident when reading the code. From the point of view of readability, I prefer to leave the explicit `IHttpResponse` declaration.

Reply:
yeah, maybe not, I wrote the code of `Execute()` so it's obvious to me  :)